### DISCUSSION TITLE
What is the best practice for combining Python and R in revealjs presentation?

### DISCUSSION THREAD

**Discussion by uriahf** [quarto-dev/quarto-cli#3408] [Q&A] [answered]
Hi all, 

I want to build a revealjs presentation that includes Python and R code. 

If I understand correctly, in R the Python code is ran by the {reticulate} package.

1. Is it possible to run R code with the Jupyter Kernel engine? IF so does it use rpy2 or something similar?
2. I can run Python code with knitr, but I cannot use some libraries such as pandas and others (though I can use numpy and random).

My main goal is to include reproducible examples in both languages, not sure what is the best way of doing this...

Thanks!

**Comment by mcanouil** [comment, marked as answer]
Both ways are "good":
- Jupyter engine + rpy2
- knitr engine + reticulate

In both cases, you need to be sure the engine is seeing the proper collection of R packages or Python libraries, i.e., you might need some extra step to help rpy2 locate your R library or to help reticulate locate your Python library.

Your issue with knitr + reticulate not able to use some your Python libraries is a setup issue.
I am no expert of reticulate, but I know you can set the location of your Python install and/or install python modules/libraries using reticulate. Both should help solve your issue.

**Comment by cscheid** [comment]
Just to clarify, all execution of code happens through a _single_ engine. If you use the jupyter engine, then you get rpy2 for R code; If you use the knitr engine, than you get reticulate for Python code.

In other words, each .qmd file is always handled by a _single_ execution engine.
